cpp/include/Ice/DefaultSliceLoader.h

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) ZeroC, Inc.
+
+#ifndef ICE_DEFAULT_SLICE_LOADER_H
+#define ICE_DEFAULT_SLICE_LOADER_H
+
+#include "SliceLoader.h"
+#include <functional>
+#include <map>
+#include <mutex>
+
+namespace IceInternal
+{
+    class DefaultSliceLoader;
+    using DefaultSliceLoaderPtr = std::shared_ptr<DefaultSliceLoader>;
+
+    /// A singleton SliceLoader that the generated code uses to register factories for classes and exceptions.
+    class ICE_API DefaultSliceLoader final : public Ice::SliceLoader
+    {
+    public:
+        /// Returns the singleton instance of DefaultSliceLoader.
+        /// @return The single instance of DefaultSliceLoader.
+        static DefaultSliceLoaderPtr instance();
+
+        virtual ~DefaultSliceLoader();
+
+        Ice::ValuePtr newClassInstance(std::string_view typeId) const final;
+        std::exception_ptr newExceptionInstance(std::string_view typeId) const final;
+
+        template<class T> void addClass(int compactId)
+        {
+            std::lock_guard lock{_mutex};
+            auto p = _classFactories.find(T::ice_staticId());
+            if (p == _classFactories.end())
+            {
+                _classFactories[T::ice_staticId()] = {[] { return std::make_shared<T>(); }, 1};
+            }
+            else
+            {
+                p->second.second++;
+            }
+
+            if (compactId >= 0)
+            {
+                _compactIdTable[compactId] = T::ice_staticId();
+
+                std::string compactIdStr{std::to_string(compactId)};
+                p = _classFactories.find(compactIdStr);
+                if (p == _classFactories.end())
+                {
+                    _classFactories[compactIdStr] = {[] { return std::make_shared<T>(); }, 1};
+                }
+                else
+                {
+                    p->second.second++;
+                }
+            }
+        }
+
+        void removeClass(std::string_view typeId, int compactId)
+        {
+            std::lock_guard lock{_mutex};
+            auto p = _classFactories.find(typeId);
+            if (p != _classFactories.end())
+            {
+                if (--p->second.second == 0)
```

Comment:
That's the same logic as in the previous FactoryTable. It's presumably to support unloading of libraries that register class/exceptions.